platforms/posix/src/main.cpp

```diff
@@ -32,656 +32,591 @@
-	     endl;
-	cout << "                   This is needed if px4 is intended to be run as a upstart job on linux" << endl;
-	cout << "<data_directory> - directory where ROMFS and posix-configs are located (if not given, CWD is used)" << endl;
-	cout << "<startup_config> - config file for starting/stopping px4 modules" << endl;
-	cout << "   -h            - help/usage information" << endl;
-}
+		DriverFramework::Framework::initialize();
 
-static void process_line(string &line, bool exit_on_fail)
-{
-	vector<string> appargs(20);
+		px4::init_once();
+		px4::init(argc, argv, "px4");
 
-	stringstream(line) >> appargs[0] >> appargs[1] >> appargs[2] >> appargs[3] >> appargs[4] >> appargs[5] >> appargs[6] >>
-			   appargs[7] >> appargs[8] >> appargs[9] >> appargs[10] >> appargs[11] >> appargs[12] >> appargs[13] >>
-			   appargs[14] >> appargs[15] >> appargs[16] >> appargs[17] >> appargs[18] >> appargs[19];
-	run_cmd(appargs, exit_on_fail);
-}
+		ret = run_startup_bash_script(commands_file, absolute_binary_path, instance);
 
-static void restore_term()
-{
-	cout << "Restoring terminal\n";
-	tcsetattr(0, TCSANOW, &orig_term);
-}
+		// We now block here until we need to exit.
+		if (pxh_off) {
+			wait_to_exit();
 
-bool px4_exit_requested(void)
-{
-	return _ExitFlag;
-}
+		} else {
+			px4_daemon::Pxh pxh;
+			pxh.run_pxh();
+		}
 
-static void set_cpu_scaling()
-{
-#if defined(__PX4_POSIX_EAGLE) || defined(__PX4_POSIX_EXCELSIOR)
-	// On Snapdragon we miss updates in sdlog2 unless all 4 CPUs are run
-	// at the maximum frequency all the time.
-	// Interestingely, cpu0 and cpu3 set the scaling for all 4 CPUs on Snapdragon.
-	system("echo performance > /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor");
-	system("echo performance > /sys/devices/system/cpu/cpu3/cpufreq/scaling_governor");
+		// When we exit, we need to stop muorb on Snapdragon.
 
-	// Alternatively we could also raise the minimum frequency to save some power,
-	// unfortunately this still lead to some drops.
-	//system("echo 1190400 > /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq");
+#ifdef __PX4_POSIX_EAGLE
+		// Sending muorb stop is needed if it is running to exit cleanly.
+		// TODO: we should check with px4_task_is_running("muorb") before stopping it.
+		std::string muorb_stop_cmd("muorb stop");
+		px4_daemon::Pxh::process_line(muorb_stop_cmd, true);
 #endif
-}
 
-#ifdef __PX4_SITL_MAIN_OVERRIDE
-int SITL_MAIN(int argc, char **argv);
+		std::string cmd("shutdown");
+		px4_daemon::Pxh::process_line(cmd, true);
 
-int SITL_MAIN(int argc, char **argv)
-#else
-int main(int argc, char **argv)
-#endif
-{
-	bool daemon_mode = false;
-	bool chroot_on = false;
+	}
 
-	tcgetattr(0, &orig_term);
-	atexit(restore_term);
+	return PX4_OK;
+}
 
-	// SIGINT
-	struct sigaction sig_int {};
-	sig_int.sa_handler = _SigIntHandler;
-	sig_int.sa_flags = 0;// not SA_RESTART!;
-	sigaction(SIGINT, &sig_int, nullptr);
+int create_symlinks_if_needed(std::string &data_path)
```

Comment:
We shouldn't be creating symlinks from main. All of this should be put in place by the install.

Reply:
We can do that as a later step, there are too many moving parts otherwise.

Reply:
Yes, that's fine I just wanted to call it out.